### PR TITLE
feat(glm): add GLM-5.3 cuDNN training support

### PR DESCRIPTION
**Pull request by HuiyingLi**
# What does this PR do ?

Add a 100-step GLM-5.3 full-parameter training recipe for the cuDNN DSA and HybridEP path, and load the released GLM-5.3 128x128 blockwise-FP8 checkpoint correctly under DTensor expert/FSDP sharding.

# Changelog

- Add a GLM-5.3 Tulu3 4K recipe using cuDNN attention, HybridEP, PP4, EP64, activation checkpointing, and 32 nodes / 256 GPUs.
- Dequantize GLM-5.3 FP8 weights before merging per-expert tensors, while preserving the BF16 tensors that do not have scale siblings.
- Add a GLM-local offset-aware Torch/Triton path that derives exact DTensor shard origins from checkpoint metadata and applies the correct FP8 blocks when a shard starts inside a 128x128 block.
- Keep the shared DeepSeek-V3 state-dict adapter unchanged so this GLM feature does not alter checkpoint behavior for other model families.
- Add GLM checkpoint-key, dequantization, scale-removal, unaligned-shard, Triton parity, and BF16 GLM-5.2 compatibility tests.

# Validation

- GLM state-dict adapter GPU tests, Slurm job `16883290`: **40 passed**.
- Real released-checkpoint offset probe, Slurm job `16883927`: a `288x6144` local shard at global row offset 288 (offset 32 within its first FP8 block) matched the global reference exactly: **exact fraction 1.0, max abs 0, relative L2 0**.
- `ruff check`, `ruff format --check`, and `git diff --check`: **passed**.
- Single-node 8-GPU, 4-layer end-to-end smoke completed forward, backward, activation checkpointing, HybridEP/cuDNN execution, and optimizer steps.
- First four layers compared against Hugging Face: logits relative L2 **0.639%**, top-1 agreement **96.61%**; representative distributed checkpoint weights were bit-exact.
- Full 32-node / 256-GPU run, Slurm job `16878255`: **100/100 steps completed**, exit code 0.
  - step 0 loss: **1.33347**
  - step 99 loss: **0.53756**
  - steps 10-99 mean loss: **0.55548**
  - steps 10-99 mean throughput: **41,664 tokens/s**
  - no non-finite metrics, FP8 scale-shape warnings, or critical rank errors
  - [Weights & Biases run](https://wandb.ai/Nemo-automodel/huiyingl_workspace/runs/glm53-tulu3-4k-cudnn-100s-3b421de0da32-16878255)

image

The distributed checkpoint/topology validation used deterministic prepacked Tulu3 4K tokens to compare against the established GLM-5.2 baseline. The checked-in recipe retains the public raw `ChatDataset` preprocessing path and enables W&B in NeMo-CI through `FINETUNE_ARGS`.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?

# Additional Information

The GLM-5.3 model architecture is unchanged from GLM-5.2; the checkpoint storage format is the relevant difference addressed here. The official GLM-5.2 checkpoint has no blockwise-FP8 configuration or scale tensors and remains on its BF16 path.

### CODE REVIEW

File: nemo_automodel/components/models/deepseek_v3/state_dict_adapter.py [hunk 1]
```diff
@@ -372,6 +436,38 @@ def _dequantize_with_triton(
     return output
 
 
+def _dequantize_with_triton_offsets(
+    weight: torch.Tensor,
+    scale_inv: torch.Tensor,
+    dtype: torch.dtype,
+    block_size: int,
+    offsets_within_first_block: tuple[int, int],
+) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (HuiyingLi) [reply nemo_automodel/components/models/deepseek_v3/state_dict_adapter.py]: Addressed in 89ffd29dd: the offset-aware implementation is now GLM-local, the shared DeepSeek-V3 adapter is unchanged, and the GLM Torch/Triton helpers document tensor shapes plus the offsets_within_first_block contract.

**Review by akoumpa — APPROVED** [review]
